man/xsindex.1
.Dd May 28, 1996
.Dt XSINDEX 1
.Os xs-httpd/3.5
.Sh NAME
.Nm xsindex
.Nd Create a simple index file for a directory
.Sh SYNOPSIS
.Nm xsindex
.Op Fl b
.Op Fl f
.Op Fl m Ar mimefile
.Op Fl s
.Op Fl t Ar mimetype
.Op Fl x Ar pattern
.Ar title
.Sh DESCRIPTION
.Ql xsindex
is a program that will create a nice index page for a
directory. What it does is make a listing of all files in
the current directory (except for special files, like
.Ql \&. ,
.Ql .xsauth ,
etc) and create a table\-like overview of them. Included
with every file is the type of that file and its size.
.Pp
The program has a number of options:
.Bl -tag -width Ds
.It Fl b
If you activate this option,
.Ql ..
will not be listed in the index.
This is useful if
.Ql ..
has no meaning in that directory.
.It Fl f
This option disabled the
.Ql Are you sure you want to overwrite?
prompt.
.It Fl m Ar mimefile
This option allows you to specify a non\-standard mime index
file. This allows you to specify, for example, different
icon images for file types.
.It Fl s
Disable the display of file sizes.
.It Fl t Ar mimetype
This option allows you to specify how file types are
displayed. The
.Ar mimetype
can be either
.Sy 1
(default) show full mime types such as
.Ql image/jpeg ,
.Sy 2
show a descriptive text such as
.Ql JPEG image
(see the
.Fl m
option) or
.Sy 3
disable display of the file type.
.It Fl x Ar pattern Oo Fl x Ar pattern ... Oc
Give a filename, or shell file\-matching pattern, for the
file(s) that should not be included in the listing. The
.Fl x
option may be included multiple times to exclude multiple
files and/or file patterns.
.El
.Pp
The
.Ar title
is used as the title of the document that will be created.
Use a descriptive title!
.Sh FILES
.Bl -tag -width Ds
.It Pa mime.index
Global configuration file containing information about
common file extensions and MIME information. Each record
(line) in this file consists of 5 tab separated fields:
.Bl -tag -offset indent -compact
.It mime type
.It file extension
.It icon file (relative to Sy HtmlDir )
.It short type name
.It full type name
.El
.It Pa index.html
HTML file that will be automatically generated by
.Xr xsindex 1 .
The contents of this file will be displayed whenever a
client requests the directory path.
.El
.Sh EXAMPLES
.Dl $ xsindex \-t 2 Qq Directory of /pub/ftp
.Sh DIAGNOSTICS
.Ex -std xsindex
.Sh SEE ALSO
.Xr httpd 1
.Pp
The project homepage:
.Pa http://www.xs\-httpd.org/


ChangeLog
This file contains an overview of version differences. The list
is not complete. Refer to cvs comments for more information.

Changes since 3.7 beta 8
- New ServerInfo option to set version string (full, branch, name)
- Pass along all HTTP headers when a forwarding proxy request

Changes since 3.7 beta 7
- Rename hier.layout to dirs.conf and shuffle configurable dirs

Changes since 3.7 beta 6
- Fix IconDir/ExecDir usage for vhosts
- Fix handling of unprivileged uid
- Fix mod_python for versions other than 2.5
- Fix mod_gzip dependency
- Fix mod_ruby initialization (enable component loading)
- Easier install directory configuration via hier.layout
- Abort configure when a selected package cannot be used

Changes since 3.7 beta 5
- Execute in .xsconf now works if script.methods is emtpy
- UseLocalScript now defaults to on (and is checked properly)
- Fix on-the-fly (de)compression: now works with large files

Changes since 3.7 beta 4
- Add UseGzipCompression to mod_gzip for on-the-fly compression
- Build mod_bzip2 by default

Changes since 3.7 beta 3
- Fix bug that caused internal: commands in .xsscripts to be ignored
- Add mod_gzip, mod_compress, mod_bzip to replace external binaries
- Add Content-MD5 trailers for CGI scripts

Changes since 3.7 beta 2
- Fix authentication module logic: all modules need to grant access

Changes since 3.7 beta 1
- Modules are now implemented as shared libraries
- Add Modules option to set which modules get loaded

Changes since 3.6 beta 24
- Fix LDAP authentication
- Introduce seperated modules (mod_ldap, perl, python, ruby)
- Make OpenSSL mandatory (removes 40 #ifdefs, 2 libcompat files)

Changes since 3.6 beta 23
- Clean up 'forward' syntax in .redir to match 'redir' command
- Include a replacement crypt() if libc, libcrypt and libssl fail
- Add NoPrivs option to .xsconf (and deprecate .xsuid)
- Enable Smashing Stack Protector (SSP) by default
- Fix parsing of empty lines in mime.types

Changes since 3.6 beta 22
- Fix possible build conflict caused by libmd and openssl
- Fix parsing of mime.types files

Changes since 3.6 beta 21
- Fix memory mismanagement in digest authentication verification
- Replace fgets(), fgetln() by friendlier fparseln() routine
  Include replacement functions for fgetln() and fparseln()
- Fix crash in https timeout handler
- Portability fixes (b20 doesn't work on many platforms)
- Tweak TCP options (especially for Linux systems)
- Add xspasswd -r option to remove accounts from password file

Changes since 3.6 beta 20
- Remove arbitrary string length limits for many configuration files,
  HTTP header messages and SSL certificate fields
- Add support for SSLVhosts (using TLS extension: Server Name Indication)
- Fix NULL-pointer dereference and string termination in HTML encoding
- Add UseSendfile option in httpd.conf to disable sendfile(2)

Changes since 3.6 beta 19
- Fix memory mismanagement in digest authentication verification

Changes since 3.6 beta 18
- Enable experimental ruby interpreter (internal:ruby)

Changes since 3.6 beta 17
- Fix parsing problem in run-cgi SSI (and maybe other places)

Changes since 3.6 beta 16
- Add index.xhtml and index.xml to default IndexFiles
- Improve url-encoding routine

Changes since 3.6 beta 15
- Fix escaping of user input used in HTML error messages;
  possible XSS abuse reported by Serge van den Boom
- Fix checks for pcre used by libxs

Changes since 3.6 beta 14
- Fix potential crash in server-side include parsing

Changes since 3.6 beta 13
- Always terminate SSL connection when reading from socket fails
- Add conditional ifenv option to .redir files

Changes since 3.6 beta 12
- Don't install empty docdir
- Shuffle internal structures (reduce global vars, getenv calls)
- Reject non-HTTP protocol specifiers

Changes since 3.6 beta 11
- Drop persistent connections after client read/write errors

Changes since 3.6 beta 10
- Add RedirFile configuration option for light-weight vhosts
- Sort qvalues in accept-* arrays
- Fake qvalues for */* and type/* when qvalues are not used

Changes since 3.6 beta 9
- Use original URL in rewriting rules before trying index.html
- Add PhIconDir option, analogous to PhExecDir

Changes since 3.6 beta 8
- Close unused CGI pipes (could block execution), by jilles

Changes since 3.6 beta 7
- Extend characters available for alternative setproctitle()
- Lock pidfile and don't start if somebody holds a lock
- Allow multiple arguments to Restrict and AuthFile
- Allow method selector in .xsconf blocks (e.g. only match POST)

Changes since 3.6 beta 6
- Improve configuration checks
- Fix include order and typechecks

Changes since 3.6 beta 5
- Enable FastCGI support again (experimental)
- Clean Perl and Python detection and initialisation (experimental)

Changes since 3.6 beta 4
- Shuffle configuration files
- Fix build outside source directory (wrt config.h)
- Fix sendfile on Linux
- Fix infinate read when content-length exceeds post data
- Portability fixes in function/library checking

Changes since 3.6 beta 3
- Remove outputbuffer and set TCP_NOPUSH (where available)
- Add 'passexist' redirection command

Changes since 3.6 beta 2
- Fix server-side include variable cleanup
- Use sendfile (where appropriate)

Changes since 3.6 beta 1
- Add http status code back to $ERROR_READABLE

Changes since 3.5 beta 31
- No functional changes

Changes since 3.5 beta 30
- Fix read errors with long POST content

Changes since 3.5 beta 29
- Make CGI header size dynamic (unlimited)
- Add default PathInfoScripts (imagemap, xschpass)
- Update manpages

Changes since 3.5 beta 28
- Fix basic authentication check
- Fix ldap authentication check

Changes since 3.5 beta 27
- Fix reading of stderr data from CGI scripts (could block CGI)
- Minor code cleanups

Changes since 3.5 beta 26
- Add -f filename option to xspasswd
- Change default of UsePut to true; document default settings
- Minor portability improvements
- Minor changes in server error messages (standard format)

src/xsindex.c
/* Copyright (C) 1995, 1996 by Sven Berkvens (sven@stack.nl) */
/* Copyright (C) 1998-2008 by Johan van Selst (johans@stack.nl) */

#include	"config.h"

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<stdbool.h>

#include	<unistd.h>
#include	<stdio.h>
#include	<stdlib.h>
#include	<string.h>
#include	<fcntl.h>
#include	<fnmatch.h>
#ifdef		HAVE_ERR_H
#include	<err.h>
#endif		/* HAVE_ERR_H */

#include	"malloc.h"
#include	"path.h"

typedef	struct	mime
{
	char		type[BUFSIZ], ext[BUFSIZ], icon[BUFSIZ],
			alt[BUFSIZ], small[BUFSIZ];
	struct	mime	*next;
} mime;

typedef struct	exlist
{
	char	*pattern;
	struct	exlist	*next;
} exlist;

static	int	show_type = 1;
static	bool	show_size = true, show_back = true, force_overwrite = false;
static	size_t	max_filename = 0, max_mimetype = 0, max_mimealt = 0,
				max_mimeshort = 0;
static	mime	*mimes;
static	const char	*mimefile;

static	void	usage			(void) NORETURN;
static	void	loadmime		(const char *);
static	const	char	*encode		(const char *) WARNUNUSED;
static	const	char	*neatsize	(off_t) WARNUNUSED;
static	const	mime	*findmime	(const char *) WARNUNUSED;

static	void
usage()
{
	fprintf(stderr,
		"Usage: xsindex [-b] [-f] [-m mimefile] [-s] [-t number] [-x pattern] title\n");
	fprintf(stderr, "   -b           Do not create a `back' link (..)\n");
	fprintf(stderr, "   -f           Do not ask whether to overwrite %s\n",
		INDEX_HTML);
	fprintf(stderr, "   -m mimefile  Use your own mime.index file\n");
	fprintf(stderr, "                (default is %s)\n", mimefile);
	fprintf(stderr, "   -s           Do not give size of each file\n");
	fprintf(stderr, "   -t number    Specify `file type' field options:\n");
	fprintf(stderr, "                   1 - Full mime type\n");
	fprintf(stderr, "                   2 - Short mime type\n");
	fprintf(stderr, "                   3 - Do not give types\n");
	fprintf(stderr, "   -x pattern   Give a filename or pattern "
		"for what should not be listed\n");
	fprintf(stderr, "   title        Title of the %s page\n", INDEX_HTML);
	fprintf(stderr, "                Use \"'s if it's more than one word\n");
	exit(1);
}

static	void
loadmime(const char *name)
{
	FILE		*input;
	char		buffer[BUFSIZ], *end;
	mime		*new;

	mimes = NULL;
	if (!(input = fopen(name, "r")))
		err(1, "fopen(`%s')", name);
	while (fgets(buffer, BUFSIZ, input))
	{
		if (strchr(buffer, '#'))
			*strchr(buffer, '#') = 0;
		end = strchr(buffer, '\0');
		while ((end > buffer) && (*(end - 1) <= ' '))
			*(--end) = 0;
		if (!buffer[0])
			continue;
		MALLOC(new, mime, 1);
		if (sscanf(buffer, "%s %s %s %s %[^\n]\n", new->type,
			new->ext, new->icon, new->alt, new->small) != 5)
		{
			fprintf(stderr, "Cannot parse line `%s'", buffer);
			continue;
		}
		new->next = mimes; mimes = new;
	}
	fclose(input);
	if (!mimes)
		errx(1, "Empty mime.index file: you need at least a default!");
}

static	const	char	*
encode(const char *what)
{
	size_t		len;
	static	char	buffer[BUFSIZ];

	buffer[0] = '\0';
	for (const char *p = what; (len = strcspn(p, "<>&\"")); p += len + 1)
	{
		if (strlen(buffer) + len < BUFSIZ)
			strncat(buffer, p, len);
		if (!p[len])
			break;
		switch (p[len])
		{
		case '<':
			strlcat(buffer, "&lt;", BUFSIZ);
			break;
		case '>':
			strlcat(buffer, "&gt;", BUFSIZ);
			break;
		case '&':
			strlcat(buffer, "&amp;", BUFSIZ);
			break;
		case '"':
			strlcat(buffer, "&quot;", BUFSIZ);
			break;
		default:
			/* do nothing */;
		}
	}
	return (buffer);
}

static	const	char	*
neatsize(off_t size)
{
	static	char	buffer1[BUFSIZ];
	char		buffer2[BUFSIZ];

	buffer1[0] = 0;
	while (size)
	{
		const off_t temp = size / 1000;
		if (temp)
			snprintf(buffer2, BUFSIZ, "%03" PRIoff ",%s",
				size % 1000, buffer1);
		else
			snprintf(buffer2, BUFSIZ, "%" PRIoff ",%s",
				size % 1000, buffer1);
		strlcpy(buffer1, buffer2, BUFSIZ);
		size = temp;
	}
	if (buffer1[0])
		buffer1[strlen(buffer1) - 1] = 0;
	else
		strlcpy(buffer1, "0", BUFSIZ);
	return (buffer1);
}

static	const	mime	*
findmime(const char *ext)
{
	const	mime	*search;
	const	char	*end;

	if (!strcmp(ext, ".."))
		end = "..";
	else if (!strcmp(ext, ".directory."))
		end = ".directory.";
	else if ((end = strrchr(ext, '.')))
		end++;
	else
		end = "txt";

	for (search = mimes; search; search = search->next)
		if (!strcasecmp(search->ext, end))
			break;
	return search ? search : mimes;
}

int
main(int argc, char **argv)
{
	int			option, amount, count;
	char			**listing, buffer[BUFSIZ];
	FILE			*output, *ls;
	struct	stat		statbuf;
	const	mime		*search;
	exlist			*exclude, *exhead;

	exclude = NULL;
	mimefile = calcpath(MIME_INDEX);
	while ((option = getopt(argc, argv, "bfm:st:x:")) != EOF)
	{
		switch(option)
		{
		case 'b':
			show_back = false;
			break;
		case 'f':
			force_overwrite = true;
			break;
		case 'm':
			mimefile = optarg;
			break;
		case 's':
			show_size = false;
			break;
		case 't':
			show_type = atoi(optarg);
			if ((show_type < 1) || (show_type > 3))
			{
				warnx("Invalid argument to -t");
				usage();
			}
			break;
		case 'x':
			MALLOC(exhead, exlist, 1);
			exhead->next = exclude;
			exhead->pattern = optarg;
			exclude = exhead;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;
	if (argc > 1)
		usage();

	loadmime(mimefile);

	if (!(ls = popen("ls -a", "r")))
		err(1, "popen(`ls -a', `r')");
	amount = 0;
	MALLOC(listing, char *, 16);
	while (fgets(buffer, BUFSIZ, ls))
	{
		bool skip = false;

		if (buffer[0] && (buffer[strlen(buffer) - 1] < ' '))
			buffer[strlen(buffer) - 1] = 0;
		if (!strcmp(buffer, ".") || !strcmp(buffer, INDEX_HTML) ||
				!strncmp(buffer, ".xs", 3) ||
				!strcmp(buffer, NOXS_FILE))
			continue;
		if ((strlen(buffer) >= 6) &&
			(!strcmp(strchr(buffer, '\0') - 6, ".redir")))
			continue;
		if (!strcmp(buffer, "..") && !show_back)
			continue;
		for (exhead = exclude; exhead; exhead = exhead->next)
			if (fnmatch(exhead->pattern, buffer, 0) != FNM_NOMATCH)
			{
				skip = true;
				printf("b %s p %s\n", buffer, exhead->pattern);
				break;
			}
		if (skip)
			continue;
		STRDUP(listing[amount], buffer);
		if (max_filename < strlen(listing[amount]))
			max_filename = strlen(listing[amount]);
		if (!((amount + 1) & 0xf))
			REALLOC(listing, char *, amount + 17);
		if (stat(listing[amount], &statbuf))
			err(1, "stat(`%s')", listing[amount]);
		if (S_ISDIR(statbuf.st_mode))
			search = findmime(".directory.");
		else
			search = findmime(listing[amount]);
		if (max_mimealt < strlen(search->alt))
			max_mimealt = strlen(search->alt);
		if (max_mimetype < strlen(search->type))
			max_mimetype = strlen(search->type);
		if (max_mimeshort < strlen(search->small))
			max_mimeshort = strlen(search->small);
		amount++;
	}
	pclose(ls);

	if (!force_overwrite && !access(INDEX_HTML, F_OK))
	{
		printf("A file called %s already exists.\n", INDEX_HTML);
		printf("Do you want to overwrite it (y/n)? ");
		fflush(stdout);
		if (!fgets(buffer, BUFSIZ, stdin) || strcmp(buffer, "y\n"))
			errx(1, "Cancelled on user's request");
	}
	remove(INDEX_HTML);
	if (!(output = fopen(INDEX_HTML, "w")))
		err(1, "fopen(%s)", INDEX_HTML);
	fprintf(output, "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
	fprintf(output, "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.1//EN\" "
		"\"http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd\">\n");
	fprintf(output, "<html><head><title>%s</title></head><body>\n",
		argc ? argv[0] : "directory listing");
	fprintf(output, "<h1>%s</h1>\n<hr />\n<pre>",
		argc ? argv[0] : "directory listing");

	for (count = 0; count < amount; count++)
	{
		if (stat(listing[count], &statbuf))
			err(1, "stat(`%s')", listing[count]);
		if (strcmp(listing[count], "..") && S_ISDIR(statbuf.st_mode))
			search = findmime(".directory.");
		else
			search = findmime(listing[count]);
		fprintf(output, "<a href=\"%s\">", encode(listing[count]));
		fprintf(output, "<img src=\"%s\" ", encode(search->icon));
		fprintf(output, "alt=\"[%s]%*.*s\" />", encode(search->alt),
			(int)(max_mimealt - strlen(search->alt)),
			(int)(max_mimealt - strlen(search->alt)), "");
		fprintf(output, "</a>  ");
		fprintf(output, "<a href=\"%s\">", encode(listing[count]));
		fprintf(output, "%s</a>%*.*s    ", encode(listing[count]),
			(int)(max_filename - strlen(listing[count])),
			(int)(max_filename - strlen(listing[count])), "");
		switch(show_type)
		{
		case 1:
			fprintf(output, "%-*.*s", (int)max_mimetype, (int)max_mimetype,
				search->type);
			break;
		case 2:
			fprintf(output, "%-*.*s", (int)max_mimeshort, (int)max_mimeshort,
				search->small);
			break;
		default:
			break;
		}
		if (show_size && (statbuf.st_mode & S_IFREG))
			fprintf(output, "   %11.11s",
				neatsize(statbuf.st_size));
		fprintf(output, "\n");
	}
	fprintf(output, "</pre></body></html>\n");
	fclose(output);
	printf("`%s' is now ready...\n", INDEX_HTML);
	return 0;
}
